### PR TITLE
CSPRNG: Support for native platform on Linux and macOS

### PR DESCRIPTION
**Pull request by kkrentz**
This PR adds functionality for seeding the CSPRNG on the native platform, at least when run on Linux or macOS.

### CODE REVIEW

File: arch/platform/native/platform.c [hunk 1]
```diff
@@ -238,7 +247,18 @@ platform_init_stage_one()
...
   leds_init();
-  return;
+#if defined(__linux__) || defined(__MACH__)
+  struct csprng_seed seed;
+#ifdef __linux__
+  if(getrandom(seed.u8, CSPRNG_SEED_LEN, GRND_RANDOM) != CSPRNG_SEED_LEN) {
+#else /* __linux__ */
+  if(SecRandomCopyBytes(kSecRandomDefault, CSPRNG_SEED_LEN, seed.u8)) {
+#endif /* __linux__ */
+    assert(false);
```
Comment (pjonsson) [inline arch/platform/native/platform.c:257]: Can this state be recovered, or is it a fatal error? If it is fatal, it sounds like this should be an abort/exit.

That would also allow eliminating the else-block and just call `csprng_feed` without indentation.

File: arch/platform/native/platform.c [hunk 2]
```diff
@@ -64,6 +64,15 @@
 #include "net/ipv6/uip-ds6.h"
 #endif /* NETSTACK_CONF_WITH_IPV6 */
 
+#include "lib/assert.h"
+#include "lib/csprng.h"
+#ifdef __linux__
+#include <sys/random.h>
+#elif defined(__MACH__)
```
Comment (pjonsson) [inline arch/platform/native/platform.c:71]: There is a single mention of `__MACH__` in the tree right now, other places use `__APPLE__` to detect compilation for macOS. Can this place use `__APPLE__` too, or is that different from `__MACH__`?

File: arch/platform/native/platform.c [hunk 3]
```diff
@@ -238,7 +247,18 @@ platform_init_stage_one()
   gpio_hal_init();
   button_hal_init();
   leds_init();
-  return;
+#if defined(__linux__) || defined(__MACH__)
+  struct csprng_seed seed;
+#ifdef __linux__
+  if(getrandom(seed.u8, CSPRNG_SEED_LEN, GRND_RANDOM) != CSPRNG_SEED_LEN) {
```
Comment (pjonsson) [inline arch/platform/native/platform.c:253]: The Windows support has been removed, and FreeBSD has getrandom, so I think the inner ifdef can be `#ifdef __APPLE__ ... #else ... #endif`, and the outer ifdef can be dropped.

File: arch/platform/native/platform.c (hunk 1 above)
Comment (kkrentz) [reply arch/platform/native/platform.c:257]: It is not that fatal as long as the return value of `csprng_rand` is always being checked. I will remove that assert.

File: arch/platform/native/platform.c [hunk 4]
```diff
@@ -238,7 +247,16 @@ platform_init_stage_one()
   gpio_hal_init();
   button_hal_init();
   leds_init();
-  return;
+#if defined(__linux__) || defined(__APPLE__)
```
Comment (pjonsson) [inline arch/platform/native/platform.c:250]: There is no Windows support, so I think this ifdef can be dropped.

File: arch/platform/native/platform.c [hunk 5]
```diff
@@ -238,7 +247,16 @@ platform_init_stage_one()
...
+  struct csprng_seed seed;
+#ifdef __linux__
```
Comment (pjonsson) [inline arch/platform/native/platform.c:252]: Make this `#ifdef __APPLE__` and flip the branches and the code should work on FreeBSD too (assuming `csprng_feed` works on FreeBSD).

**Review by nvt — APPROVED** [review]

**Review by nfi — APPROVED** [review]

**Pull request merged**